### DISCUSSION TITLE
Group feature selection

### DISCUSSION THREAD

**Discussion by kerrymcgowan** [rasbt/mlxtend#954] [Q&A]
Hello Sebastian,

Is it possible to perform feature selection with groups of features instead of individual features with mlxtend? So for example, if you have 100 features that could be partitioned into groups 1-10, is there a way to run feature selection to select which combination of those 10 groups produces the best model instead of selecting from the 100 features individually?

**Comment by rasbt** [comment]
Good question! The answer is yes and no 😅. Right now, it doesn't work out of the box, but I remember doing something similar once where I had 3 features, I did a one-hot encoding on the third feature, which became 20 new features due to the one hot encoding, leaving me with 22 features in total. I wrote some custom code / modification so that the the 20 new features were treated as a group rather than 20 individual ones.

I should have that code somewhere and will try to see if I can find it in the next couple of days. I should still have it somewhere since that was for a research project 1-2 years ago that I should have archived.

**Reply by NimaSarajpoor** [reply]
@rasbt 

If  you do not mind, I am going to share my thoughts. I was JUST working on "Credit" data to reproduce a figure provided in the  book "Introduction to Statistical Learning", and I realized one of the categorical feature is not binary, which means I may need to transform it to more  than one column (there are some ways to convert it to numerical column and keep it still one column). 

I took a look at "https://github.com/rasbt/mlxtend/blob/master/mlxtend/feature_selection/exhaustive_feature_selector.py"

and I think it does not support a situation where a column is transformed to more than one column. right? That is why you did:

> I wrote some custom code / modification so that the the 20 new features were treated as a group rather than 20 individual ones.

----

So, I think it would be nice to add a parameter, like `is_together`, which can be 2D array of type `bool`, and with shape `(n_feature, n_feature)`, and then if `is_together[i. j] = True`, then we should consider feature `i` and `j` together. The default of this new parameter can be diagonal matrix and all elements on the diagonal is True and, all other elements are False.

What do you think?

----

Update:
Apparently, the authors of the book consider each of the new encoded feature on its own and did not group them together in their exhaustive feature selection. However, I still think having that feature in mlxtend would be cool :)

**Reply by rasbt** [reply]
Arg, sorry. Actually, I misremembered and the code was for the feature importance analysis. I think for the SFS it would be a bit more work. I am not sure if I currently have the time for that.

**Reply by NimaSarajpoor** [reply]
@rasbt 
Is that okay if I start working on this? do you have time to help throughout the process? If you think it is too much and probably not useful for most users, please feel free to let me know :)

**Reply by rasbt** [reply]
Absolutely! That would be amazing. I am happy to try to help along the way when I can

**Reply by rasbt** [reply]
Btw I think it can probably be implemented similar to the `feature_permutation_importance` (https://github.com/rasbt/mlxtend/blob/master/mlxtend/evaluate/feature_importance.py#L116, this can be refactored, but currently no time, just wanted to get a correctly working base version done, and refactoring is a future problem 😅)

**Reply by NimaSarajpoor** [reply]
> Btw I think it can probably be implemented similar to the `feature_permutation_importance`

Cool. Thanks for letting me know. I will take a look!

**Reply by kerrymcgowan** [reply]
Thank you for checking @rasbt. For anyone looking into doing something similar, I was able to find a group forward feature selection tutorial using mlxtend and eobox [here](https://eo-box.readthedocs.io/en/latest/examples/ml/ffgs_ForwardFeatureGroupSelection.html).
